staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -572,8 +574,29 @@ func (w *watchCache) Replace(objs []interface{}, resourceVersion string) error {
 	w.Lock()
 	defer w.Unlock()
 
-	w.startIndex = 0
-	w.endIndex = 0
+	// We have to ensure that startIndex is never decreased.
+	// Previously we were resetting startIndex to 0 and that was exposing
+	// us to the following race:
+	// - a new Watch was registered, and watchCacheInterval is returned for it
+	//   (but it didn't yet started processing the interval)
+	// - Cacher is being reinitialized (which terminates this watcher, but
+	//   that only prevents delivering new events to it, but the ones that
+	//   are already queued may still be delievered to the watch)
+	// - watchCache is receiving new events which start to being written
+	//   starting at index 0
+	// - now the Watch from the first point starts actual processing - it
+	//   tries to load the data, but if it was conflicting with the already
+	//   overwritten indexes, it will return bad data (from the future
+	//   which then may be followed by past data for indexes which weren't
+	//   yet overwritten).
+	//
+	// Ensuring that startIndex is not decreasing, prevents us from this
+	// scenario thanks to the indexValidator which checks if the index
+	// the interval wants to download is still a valid one.
```

Comment:
I would probably simplify this whole comment to:
```
Ensure startIndex never decreases, so that existing watchCacheInterval instances get invalid errors
if they try to download from the buffer using their own start/end indexes calculated from previous buffer content.
```

Reply:
done